File: 0133-clone-graph/0133-clone-graph.js
/**
 * // Definition for a _Node.
 * function _Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {_Node} node
 * @return {_Node}
 */
var cloneGraph = function(node) {
    if(!node){
        return null
    }
    const map = new Map();

    function copyNode(n){
        if(map.has(n)){
            return map.get(n);
        }
        let copy = new Node(n.val);
        map.set(n, copy);
        copy.neighbors = n.neighbors.map(copyNode);
        
        return copy;
    }

    return copyNode(node);
};

// base case
// if no node, return null

// map to store copied node
// function copyNode
// if map has node, return map.get node
// if map doesn't have node
// create a new node, add the new node to map
// get the added node, and copy neighbors,
// node.neighbors.map(copy)
// return new node
// return copyNode(node)

File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if node is None:
            return None

        visited = {}

        def helper(n):
            if n in visited:
                return visited[n]

            if n not in visited:
                copy = Node(n.val)
                visited[n] = copy

                for neighbor in n.neighbors:
                    copy.neighbors.append(helper(neighbor))
                
                return copy
        
        return helper(node)

        